## Import libraries

In [2]:
# Import analytical libraries
import pandas as pd
import numpy as np
import os

## Import Data

In [7]:
# Create Path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [8]:
# Import Customers dataframe
df_customers=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col=False)

In [9]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
df_customers.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [13]:
# Check for the number of rows and columns
df_customers.shape

(206209, 10)

In [14]:
# Check the type of data
c.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [18]:
# Descriptive statistics on Customers dataframe
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

### Observations: The basic criteria of descriptive statistics do not show significant deviations in data values. There is a significant range of values in the income column. We need more information about how the income data was collected. For instance, if customers have selected their income from the options available on the app, there should not be an error. If they have entered the income manually, the error is possible. We will check for outliers.  

## Wrangling Customers dataframe

In [27]:
# Create a dictionary of old and new columns name
customers_columns_map={
    'First Name': 'first_name',
    'Surnam': 'last_name',
    'STATE': 'states',
    'Age': 'age',
    'n_dependants': 'number_dependants',
    'fam_status': 'family_status'
}

In [28]:
# Rename columns using the columns dictionary
df_customers.rename(columns=customers_columns_map, inplace=True)

In [29]:
# Check for Customers data after renaming columns
df_customers.head()

user_id first_name last_name  Gender      states  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_dependants family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

In [30]:
df_customers.shape

(206209, 10)

In [31]:
# Rename column 'Gender'
df_customers.rename(columns={'Gender': 'gender'}, inplace=True)

In [32]:
df_customers.head()

user_id first_name last_name  gender      states  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_dependants family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

### Observations: I have renamed the columns to get the columns titles normalized to the same format and relevant to the columns content. 

In [34]:
# Check for the missing values on Customers dataframe
df_customers.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
states                   0
age                      0
date_joined              0
number_dependants        0
family_status            0
income                   0
dtype: int64

In [36]:
# Check for NaN values on Customers dataframe
df_customers.isna().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
states                   0
age                      0
date_joined              0
number_dependants        0
family_status            0
income                   0
dtype: int64

### Observations: There are 11,259 missing values on the first_name column. We don’t have reliable data to replace missing values. Considering he objective of this project of analyzing the customers behavior and the key questions of the stakeholders, we can drop the columns with first and last name.

In [46]:
# Remove columns 'first_name' and 'last_name' from Customers dataframe
df_customers=df_customers.drop(columns=['first_name', 'last_name'])

In [47]:
# Check for Customers dataframe after dropping the columns
df_customers.head()

user_id  gender      states  age date_joined  number_dependants  \
0    26711  Female    Missouri   48    1/1/2017                  3   
1    33890  Female  New Mexico   36    1/1/2017                  0   
2    65803    Male       Idaho   35    1/1/2017                  2   
3   125935  Female        Iowa   40    1/1/2017                  0   
4   130797  Female    Maryland   26    1/1/2017                  1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

## Data Consitency Checks

In [48]:
# Create a new dataframe containing duplicate rows on Customers 
df_customers_dup=df_customers[df_customers.duplicated()]

In [49]:
df_customers_dup

Empty DataFrame
Columns: [user_id, gender, states, age, date_joined, number_dependants, family_status, income]
Index: []

### Observations: There are no duplicates on Customers dataframe.

In [50]:
# Check for mix type data
for col in df_customers.columns.tolist():
    weird=(df_customers[[col]].map(type) !=df_customers[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_customers [weird])>0:
        print(col)

### Observations: no mixed-type data

In [53]:
# Check for the number of rows and columns after cleaning the Customers data
df_customers.shape

(206209, 8)

## Combining Orders-Products and Customers Dataframes

In [55]:
# Import Orders-Product dataframe
df_ords_prods_agg=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [56]:
df_ords_prods_agg.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ...    price_range_loc   busiest_day  busiest_slowest_days  \
0          0  ...  Mid range product  Regular busy          Regular busy   
1          1  ...  Mid range product  Regular busy          Slowest days   
2          1  ...  Mid range product  Regular busy          Slowest days   
3          1  ...  Mid range product    Least busy          Slowest days   
4          1  ...  Mid range product    Least busy          Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spending  \
0         Average orders        10  New customer         6.367797   
1         Average orders        10  New customer         6.367797   
2            Most orders        10  New customer         6.367797   
3         Average orders        10  New customer         6.367797   
4            Most orders        10  New customer         6.367797   

  spender_flag  median_days_prior_order   order_frequency_flag  
0  Low spender                     20.5  Non-frequent customer  
1  Low spender                     20.5  Non-frequent customer  
2  Low spender                     20.5  Non-frequent customer  
3  Low spender                     20.5  Non-frequent customer  
4  Low spender                     20.5  Non-frequent customer  

[5 rows x 24 columns]

In [57]:
df_ords_prods_agg.shape

(32399732, 24)

In [58]:
# Check data type on Orders - Products dataframe
df_ords_prods_agg.dtypes

order_id                     int32
user_id                      int32
order_number                 int16
orders_day_of_week            int8
order_hour_of_day             int8
days_since_prior_order     float64
first_order                   bool
product_id                   int32
add_to_cart_order            int16
reordered                     int8
product_name                object
aisle_id                     int16
department_id                 int8
prices                     float64
price_range_loc             object
busiest_day                 object
busiest_slowest_days        object
busiest_period_of_day       object
max_order                    int16
loyalty_flag                object
average_spending           float64
spender_flag                object
median_days_prior_order    float64
order_frequency_flag        object
dtype: object

In [59]:
# Check data type of Customers dataframe
df_customers.dtypes

user_id               int64
gender               object
states               object
age                   int64
date_joined          object
number_dependants     int64
family_status        object
income                int64
dtype: object

In [62]:
# Check min and max values of user_id on Oders-Products 
df_ords_prods_agg['user_id'].describe()

count    3.239973e+07
mean     1.029367e+05
std      5.946623e+04
min      1.000000e+00
25%      5.142000e+04
50%      1.026140e+05
75%      1.543880e+05
max      2.062090e+05
Name: user_id, dtype: float64

In [63]:
# Check min and max values of user_id on Customers 
df_customers['user_id'].describe()

count    206209.000000
mean     103105.000000
std       59527.555167
min           1.000000
25%       51553.000000
50%      103105.000000
75%      154657.000000
max      206209.000000
Name: user_id, dtype: float64

### Observations: We will merge two dataframes on uder_id using inner join. The range of values of user_id from 1 to 206,209 is the same on both dataframes. Before merging we will change the data type of user_id column of Customers for ‘int32’ to get the same data type of user_id values in both dataframes.  In order to reduce memory usage when combining files the datatype of two other columns on Customers will be changed as well. i.e. ‘age’ and ‘number_dependants’ => ‘int8’ 

In [66]:
# Optimize data type of 'user_id', 'age', 'nember_dependants' on Customer dataframe
df_customers_optim=df_customers.astype({'user_id': 'int32', 'age': 'int8', 'number_dependants': 'int8'})

In [67]:
# Check the data type on optimized Customers dataframe
df_customers_optim.dtypes

user_id               int32
gender               object
states               object
age                    int8
date_joined          object
number_dependants      int8
family_status        object
income                int64
dtype: object

In [68]:
df_customers_optim.head()

user_id  gender      states  age date_joined  number_dependants  \
0    26711  Female    Missouri   48    1/1/2017                  3   
1    33890  Female  New Mexico   36    1/1/2017                  0   
2    65803    Male       Idaho   35    1/1/2017                  2   
3   125935  Female        Iowa   40    1/1/2017                  0   
4   130797  Female    Maryland   26    1/1/2017                  1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [69]:
df_customers_optim.describe()

user_id            age  number_dependants         income
count  206209.000000  206209.000000      206209.000000  206209.000000
mean   103105.000000      49.501646           1.499823   94632.852548
std     59527.555167      18.480962           1.118433   42473.786988
min         1.000000      18.000000           0.000000   25903.000000
25%     51553.000000      33.000000           0.000000   59874.000000
50%    103105.000000      49.000000           1.000000   93547.000000
75%    154657.000000      66.000000           3.000000  124244.000000
max    206209.000000      81.000000           3.000000  593901.000000

## Combining data

In [72]:
# Merge Orders-Products and Customers dataframe
df_instacart_merged=df_ords_prods_agg.merge(df_customers_optim, on='user_id', indicator=True)

In [74]:
df_instacart_merged

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32399727    156685   106143            26                   4   
32399728    484769    66343             1                   6   
32399729   1561557    66343             2                   1   
32399730    276317    66343             3                   6   
32399731   2922475    66343             4                   1   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         7                    15.0        False         196   
2                        12                    21.0        False         196   
3                         7                    29.0        False         196   
4                        15                    28.0        False         196   
...                     ...                     ...          ...         ...   
32399727                 23                     5.0        False       19675   
32399728                 11                     NaN         True       47210   
32399729                 11                    30.0        False       47210   
32399730                 15                    19.0        False       47210   
32399731                 12                    30.0        False       47210   

          add_to_cart_order  reordered  ... median_days_prior_order  \
0                         1          0  ...                    20.5   
1                         1          1  ...                    20.5   
2                         1          1  ...                    20.5   
3                         1          1  ...                    20.5   
4                         1          1  ...                    20.5   
...                     ...        ...  ...                     ...   
32399727                  1          1  ...                     7.0   
32399728                  1          0  ...                    30.0   
32399729                  1          1  ...                    30.0   
32399730                  1          1  ...                    30.0   
32399731                  1          1  ...                    30.0   

           order_frequency_flag  gender     states age date_joined  \
0         Non-frequent customer  Female    Alabama  31   2/17/2019   
1         Non-frequent customer  Female    Alabama  31   2/17/2019   
2         Non-frequent customer  Female    Alabama  31   2/17/2019   
3         Non-frequent customer  Female    Alabama  31   2/17/2019   
4         Non-frequent customer  Female    Alabama  31   2/17/2019   
...                         ...     ...        ...  ..         ...   
32399727      Frequent customer    Male     Hawaii  25   5/26/2017   
32399728  Non-frequent customer  Female  Tennessee  22   9/12/2017   
32399729  Non-frequent customer  Female  Tennessee  22   9/12/2017   
32399730  Non-frequent customer  Female  Tennessee  22   9/12/2017   
32399731  Non-frequent customer  Female  Tennessee  22   9/12/2017   

         number_dependants family_status  income _merge  
0                        3       married   40423   both  
1                        3       married   40423   both  
2                        3       married   40423   both  
3                        3       married   40423   both  
4                        3       married   40423   both  
...                    ...           ...     ...    ...  
32399727                 0        single   53755   both  
32399728                 3       married   46151   both  
32399729                 3       married   46151   b

In [76]:
# Check the merge indicator on the combuned dataframe
df_instacart_merged['_merge'].value_counts()

_merge
both          32399732
left_only            0
right_only           0
Name: count, dtype: int64

In [77]:
df_instacart_merged.shape

(32399732, 32)

In [78]:
# Drop merge flag column before extracting the data
df_instacart_merged_drop=df_instacart_merged.drop(columns=['_merge'])

In [79]:
df_instacart_merged_drop.shape

(32399732, 31)

In [80]:
df_instacart_merged_drop.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ... spender_flag  median_days_prior_order  \
0          0  ...  Low spender                     20.5   
1          1  ...  Low spender                     20.5   
2          1  ...  Low spender                     20.5   
3          1  ...  Low spender                     20.5   
4          1  ...  Low spender                     20.5   

    order_frequency_flag  gender   states age date_joined number_dependants  \
0  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
1  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
2  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
3  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   
4  Non-frequent customer  Female  Alabama  31   2/17/2019                 3   

   family_status income  
0        married  40423  
1        married  40423  
2        married  40423  
3        married  40423  
4        married  40423  

[5 rows x 31 columns]

## Expoting data

In [83]:
# Export combined Orders-Products-Customers dataframe
df_instacart_merged_drop.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'instacart_merged.pkl'))